### ISSUE TITLE
L-Building with gabled roofs cause problems

### ISSUE THREAD

**Issue by Matthias84** [tordanik/OSM2World#14]
Currently O2W has problems with buildings that have an L-groundshape and that have gabled roofs.

Here an example:
http://www.openstreetmap.org/?lat=54.07847&lon=12.114041&zoom=18 (Waldorf Kindergarten)
![Bildschirmfoto vom 2013-01-12 20:17:23](https://f.cloud.github.com/assets/217511/62221/12cf0860-5ced-11e2-83f1-b4e35a1ccf3a.png)
Even if the desired model can be enforced using a single building:part for one part of the L, this won't fix the diagonal connection between both shapes in the inner corner of both parts.
![Bildschirmfoto vom 2013-01-12 20:19:45](https://f.cloud.github.com/assets/217511/62222/17707890-5ced-11e2-9b19-8b93ce5acdf5.png)

**Comment by The-Nickname** [comment]
Hey I noticed [OSM2World is making progress](https://osm2world.org/blog) and just wanted to chime in and highlight how things are regarding this issue, because I view this issue as very annoying in OSM2World and I hope my post helps.
L-, U- or T- or H-Shaped roofs are still rendered incorrectly by OSM2World, see for example: 
<details>

<summary> https://www.openstreetmap.org/way/83694993 </summary>
<img width="2007" height="1415" alt="OSM2World" src="https://github.com/user-attachments/assets/eefb0ef3-4807-4976-9908-a87baf32f6cc" />
This is [F4map](https://demo.f4map.com) for comparison, it is rendered correctly although a bit messy: 
<img width="1850" height="1442" alt="F4map Demo" src="https://github.com/user-attachments/assets/fa570ee6-a3b7-42be-aed8-2528811a4ac6" />

This is [StreetsGL](https://github.com/StrandedKitty/streets-gl) for comparison, although the data is from 2023 in this screenshot: 
<img width="1539" height="1188" alt="StreetsGL" src="https://github.com/user-attachments/assets/5b4ec809-971f-4354-9302-759d577dcd36" />

While this is how it looks [IRL](https://atlas.bayern.de/?redirect=true&r=0&c=691366%2C5335085&mid=1&z=12&l=vt_standard): 
<img width="847" height="652" alt="IRL image" src="https://github.com/user-attachments/assets/6164ae04-a30b-4b32-81d8-39960848e404" />
</details>

<details>

<summary> https://www.openstreetmap.org/way/85513547 and https://www.openstreetmap.org/way/85513530 </summary>

<img width="1886" height="1315" alt="OSM2World" src="https://github.com/user-attachments/assets/dab54927-04aa-43ba-80f6-6e857dcb7522" />

This is [F4map](https://demo.f4map.com) for comparison, it is rendered correctly: 
<img width="2105" height="1753" alt="F4map" src="https://github.com/user-attachments/assets/fc682f40-033e-46f5-aa84-749eda5d1ba5" />

This is [StreetsGL](https://github.com/StrandedKitty/streets-gl), although the data is also from 2023: 
<img width="1669" height="1255" alt="StreetsGL" src="https://github.com/user-attachments/assets/157fdaac-39a1-450a-ab63-e8caef3c4787" />

And for comparison again the real areal images: 
<img width="1034" height="886" alt="IRL image" src="https://github.com/user-attachments/assets/254503d2-e62f-41ab-9a5c-680891d3ca8d" />
</details>

So in conclusion L-, U- or T- or H-Shaped roofs, which there are a lot of in basically every city, are often rendered incorrectly in OSM2World. While I feel that F4map and StreetsGL are more often than not correct in their display of roofs. For OSM2World it kinda feels to be the other way around, which is sad since F4map is proprietary and StreetsGL is kinda stale and not as easy to setup as OSM2World. Also this issue has been present for 13 years in OSM2World.

**Comment by tordanik** [comment]
> L-, U- or T- or H-Shaped roofs are still rendered incorrectly by OSM2World

I've recently looked into the issue of non-straight ridges again and managed to build a first working implementation:

<img width="500" alt="Ridge with corners in OSM2World" src="https://github.com/user-attachments/assets/f381ac43-ad6c-460b-9d9f-d69ce389853e" />

The main obstacle at the moment is that the Straight Skeleton libraries have tried (such as grassfire4j) have some dependencies which don't work with TeaVM, so I'd likely have to fork them.

There are also still some [unresolved tagging questions](https://community.openstreetmap.org/t/what-should-trigger-3d-roofs-with-complex-ridges/134915) because roofs with non-straight ridges aren't really standard Simple 3D Buildings mapping, but if I get the feature to work, I'd likely use some heuristics as I suggested in that thread.
